Replacing the $where operator by $expr

anupamme opened the pull request:
This is for the issue #4148

fgalan commented:
I have done a test pass with -noCache enabled and this is the result:

```
$ CB_MAX_TRIES=1 ./testHarness.sh --noCache
...

================================================================================

----------- Failing tests ------------------
  o  0000_ngsi10_subscriptions/subscription_sequence_wildcards_onchange_append-simpler.test
  o  0000_ngsi10_subscriptions/subscription_sequence_wildcards_onchange_append.test
  o  0000_ngsi10_subscriptions/subscription_sequence_wildcards_onchange_update.test
  o  0000_subs_attrs_deletion/delete_attribute_in_condition.test
  o  0000_subs_attrs_deletion/delete_attribute_in_empty_condition.test
  o  0001_federation/federation_location_ngsiv1.test
  o  0001_federation/federation_location_ngsiv2.test
  o  0674_servicepath_in_notifications/service_path_in_onchange_notification.test
  o  0674_servicepath_in_notifications/service_path_in_single_onchange_notification.test
  o  0675_servicePath_in_ngsi10_subscriptions/simple_ngsi10_subscription_both_cities_with_isPattern_and_deeper_service_path.test
  o  0675_servicePath_in_ngsi10_subscriptions/simple_ngsi10_subscription_exact_city_with_isPattern_and_deeper_service_path.test
  o  0675_servicePath_in_ngsi10_subscriptions/simple_ngsi10_subscription_with_isPattern_and_deeper_service_path.test
  o  0675_servicePath_in_ngsi10_subscriptions/simple_ngsi10_subscription_with_recursive_service_path_and_isPattern.test
  o  0706_direct_https_notifications/direct_https_notifications.test
  o  0706_direct_https_notifications/direct_https_notifications_no_accept_selfsigned.test
  o  0876_attribute_dates/attribute_dates.test
  o  0876_attribute_dates/attrs_dates_overridden_by_user_subs.test
  o  0876_attribute_dates/filter_attr_datecreated_sub.test
  o  0876_attribute_dates/filter_attr_datemodified_sub.test
  o  0876_entity_dates/entity_dates_overridden_by_user_subs.test
  o  0876_entity_dates/filter_datecreated_sub.test
  o  0876_entity_dates/filter_datemodified_sub.test
  o  0943_overnotification_bug/overnotification_bug.test
  o  0978_filter_attribute_pattern/csub_attribute_pattern.test
  o  0978_filter_attribute_pattern/csub_attribute_pattern_nocache.test
  o  1048_subscription_cache/csub_update_with_notif.test
  o  1129_strings_in_q_uri_param/q_query_strings_for_subscriptions.test
  o  1156_mq_as_uri_param_for_metadata/mq_as_uri_param_for_metadata.test
  o  1156_mq_as_uri_param_for_metadata/q_and_mq_as_uri_param_for_metadata.test
  o  1156_mqfilters_and_compounds/mqfilters_and_compounds_deeper.test
  o  1156_mqfilters_and_compounds/mqfilters_and_compounds_equals_subs.test
  o  1156_mqfilters_and_compounds/mqfilters_and_compounds_existence_subs.test
  o  1156_mqfilters_and_compounds/mqfilters_and_compounds_lt_le_gt_ge_subs.test
  o  1156_mqfilters_and_compounds/mqfilters_and_compounds_patternmatch_subs.test
  o  1156_qfilters_and_compounds/qfilters_and_compounds_deeper.test
  o  1156_qfilters_and_compounds/qfilters_and_compounds_equals_subs.test
  o  1156_qfilters_and_compounds/qfilters_and_compounds_existence_subs.test
  o  1156_qfilters_and_compounds/qfilters_and_compounds_lt_le_gt_ge_subs.test
  o  1156_qfilters_and_compounds/qfilters_and_compounds_patternmatch_subs.test
  o  1308_subscription_cache/cache_reset_from_db.test
  o  1308_subscription_cache/subCacheThrottlingAndExpiration.test
  o  1308_subscription_cache/subCacheThrottlingAndExpirationAfterSubUpdate.test
  o  1308_subscription_cache/subCacheThrottlingAndExpirationAfterSubUpdateWhenSubExpired.test
  o  1316_create_subscription/post_subs_expr_q_exist_attr.test
  o  1316_create_subscription/post_subs_expr_q_not_exist_attr.test
  o  1328_pause_subscriptions/pause_subscriptions_notif_cache.test
  o  1328_pause_subscriptions/pause_subscriptions_notif_nocache.test
  o  1449_notification_mode_cases/onchange_persisent.test
  o  1449_notification_mode_cases/onchange_simulated.test
  o  1449_notification_mode_cases/onchange_threadpool.test
  o  1494_subscription_alteration_types/sub_alteration_type_entity_change.test
  o  1494_subscription_alteration_types/sub_alteration_type_entity_create.test
  o  1494_subscription_alteration_types/sub_alteration_type_entity_delete.test
  o  1494_subscription_alteration_types/sub_alteration_type_entity_delete2.test
  o  1494_subscription_alteration_types/sub_alteration_type_entity_update.test
  o  1494_subscription_alteration_types/sub_alteration_type_entity_update3.test
  o  1678_geosubscriptions/geosubscription_mindistance_zero.test
  o  1678_geosubscriptions/geosubscriptions.test
  o  1678_geosubscriptions/geosubscriptions_plus_q_filter.test
  o  1705_csub_cache_objects/csub_cache_string_filter_update.test
  o  1898_no_multiservice_and_tenant_used/no_multiservice_and_tenant_used.test
  o  1903_colon_as_double_equal_synonym/csub_filter_colon_as_enum.test
  o  1903_colon_as_double_equal_synonym/csub_filter_colon_as_equal.test
  o  1903_colon_as_double_equal_synonym/csub_filter_colon_as_range.test
  o  1945_filter_string_comparison/csub_filter_string_greater_or_equal_than.test
  o  1945_filter_string_comparison/csub_filter_string_greater_than.test
  o  1945_filter_string_comparison/csub_filter_string_inside_range.test
  o  1945_filter_string_comparison/csub_filter_string_less_or_equal_than.test
  o  1945_filter_string_comparison/csub_filter_string_less_than.test
  o  1945_filter_string_comparison/csub_filter_string_outside_range.test
  o  2015_notification_templates/notification_templates_many_notifications_after_update.test
  o  2015_notification_templates/notification_templates_many_queue_statistics.test
  o  2064_attrs_blacklist_subs/2064_attrs_blacklist_complex.test
  o  2064_attrs_blacklist_subs/blacklist_after_sub_up.test
  o  2064_attrs_blacklist_subs/whitelist_after_sub_up.test
  o  2222_notif_not_sent_on_true/notif_not_sent_on_true.test
  o  2445_mq_date_processing_subs/mq_date_processing_subs.test
  o  2507_csub_metadata_field/special_metadata_in_notifications.test
  o  2507_csub_metadata_field/special_metadata_in_notifications_ngsiv1.test
  o  2549_append_or_update_in_actionType_in_notifications/append_or_update_in_actionType_in_notifications.test
  o  2553_wrong_metadata_type_in_prev_value_in_notification/wrong_metadata_type_in_prev_value_in_notification.test
  o  2588_missing_prevval_on_changed_type/2588_missing_prevval_on_changed_type.test
  o  2670_milliseconds_for_DateTime_attrs/2670_milliseconds_filter_sub.test
  o  2750_common_metrics/2750_notifications.test
  o  2986_apiv2_notifications/apiv2_notification_all_json_types.test
  o  2986_apiv2_notifications/apiv2_notification_common_geo_json_types.test
  o  2986_apiv2_notifications/apiv2_notification_reception.test
  o  2995_q_in_subscriptions/mq_compound_does-not-exist_sub_GTE.test
  o  2995_q_in_subscriptions/mq_compound_does-not-exist_sub_LTE.test
  o  2995_q_in_subscriptions/mq_compound_non-compatible-type_sub_GTE.test
  o  2995_q_in_subscriptions/mq_compound_non-compatible-type_sub_LTE.test
  o  2995_q_in_subscriptions/mq_value_does-not-exist_sub_GTE.test
  o  2995_q_in_subscriptions/mq_value_does-not-exist_sub_LTE.test
  o  2995_q_in_subscriptions/mq_value_non-compatible_sub_GTE.test
  o  2995_q_in_subscriptions/mq_value_non-compatible_sub_LTE.test
  o  2995_q_in_subscriptions/q_compound_does-not-exist_sub_GTE.test
  o  2995_q_in_subscriptions/q_compound_does-not-exist_sub_LTE.test
  o  2995_q_in_subscriptions/q_compound_non-compatible-type_sub_GTE.test
  o  2995_q_in_subscriptions/q_compound_non-compatible-type_sub_LTE.test
  o  2995_q_in_subscriptions/q_value_does-not-exist_sub_GTE.test
  o  2995_q_in_subscriptions/q_value_does-not-exist_sub_LTE.test
  o  2995_q_in_subscriptions/q_value_non-compatible_sub_GTE.test
  o  2995_q_in_subscriptions/q_value_non-compatible_sub_LTE.test
  o  2998_null_not_working_in_q_and_mq_filter/null_not_working_in_mq_for_subscription.test
  o  2998_null_not_working_in_q_and_mq_filter/null_not_working_in_q_for_subscription.test
  o  3001_mqtt/mqtt_notification_custom.test
  o  3001_mqtt/mqtt_subscription_multibroker.test
  o  3154_notification_problem_with_custom_header/notification_problem_with_custom_header.test
  o  3310_preupdate_metadata_content_included_in_notifications/preupdate_metadata_content_included_in_notifications.test
  o  3533_null_in_datetime_and_geo/null_in_datetime_attr_subs.test
  o  3533_null_in_datetime_and_geo/null_in_datetime_md_subs.test
  o  3533_null_in_datetime_and_geo/null_in_geo_box_subs.test
  o  3533_null_in_datetime_and_geo/null_in_geo_json_subs.test
  o  3533_null_in_datetime_and_geo/null_in_geo_line_subs.test
  o  3533_null_in_datetime_and_geo/null_in_geo_point_subs.test
  o  3533_null_in_datetime_and_geo/null_in_geo_polygon_subs.test
  o  3906_onlychangedattrs_not_working_with_csub_cache/onlychangedattrs_not_working_with_csub_cache.test
  o  4069_status_inactive_not_working_with_cache_in_ha/status_inactive_not_working_with_cache_in_ha.test
  o  4094_posix_regex/posix_regex_idpattern.test
  o  4120_string_and_null_mixed_filter/sub_string_and_null_mixed_filter.test
```

A huge number of tests are failing... I have had a look to the detail of a couple of them:

![imagen](https://user-images.githubusercontent.com/1534240/190465214-6c5f35b5-ebde-45af-9cd3-9f8b6ffb15c5.png)

![imagen](https://user-images.githubusercontent.com/1534240/190465256-60518389-318e-49e4-ba22-26753490a932.png)

So it seems that expected notification are not generated. Probably this is due to the new query has some flaw and it doesn't "capture" the subscriptions associated to the notification that should be triggered according to test script.

Just in case this feedback may be useful...

fgalan commented:
I have also done a test pass with -cache enabled and this is the result:

```
$ CB_MAX_TRIES=1 ./testHarness.sh --cache

...
---------- Failing tests ------------------
  o  0978_filter_attribute_pattern/csub_attribute_pattern_nocache.test
  o  1328_pause_subscriptions/pause_subscriptions_notif_nocache.test
  o  1494_subscription_alteration_types/sub_alteration_type_entity_update3.test
  o  3001_mqtt/mqtt_subscription_multibroker.test
```

(Let's ignore mqtt_subscription_multibroker.test... I think my MQTT broker is not working properly, so I'm unsure about this is an actual fail)

In the first two cases, the broker is started using `brokerStart CB 0 IPv4 -noCache`, which overrides the `--cache` setting at testHarness.sh level. So it makes they are failing, and this happens in the same way I described in my previous comment (i.e. notification not being sent).

Howeve, in the third case (sub_alteration_type_entity_update3) Orion is not started with `-noCache` but the problem seems to be the same (notification not being sent). I cannot explain at this moment why this test is failing...

Just in case this feedback may be useful...

anupamme commented:
> I have done a test pass with -noCache enabled and this is the result:
> 
> ```
> $ CB_MAX_TRIES=1 ./testHarness.sh --noCache
> ...
> 
> ================================================================================
> 
> ----------- Failing tests ------------------
>   o  0000_ngsi10_subscriptions/subscription_sequence_wildcards_onchange_append-simpler.test
>   o  0000_ngsi10_subscriptions/subscription_sequence_wildcards_onchange_append.test
>   o  0000_ngsi10_subscriptions/subscription_sequence_wildcards_onchange_update.test
>   o  0000_subs_attrs_deletion/delete_attribute_in_condition.test
>   o  0000_subs_attrs_deletion/delete_attribute_in_empty_condition.test
>   o  0001_federation/federation_location_ngsiv1.test
>   o  0001_federation/federation_location_ngsiv2.test
>   o  0674_servicepath_in_notifications/service_path_in_onchange_notification.test
>   o  0674_servicepath_in_notifications/service_path_in_single_onchange_notification.test
>   o  0675_servicePath_in_ngsi10_subscriptions/simple_ngsi10_subscription_both_cities_with_isPattern_and_deeper_service_path.test
>   o  0675_servicePath_in_ngsi10_subscriptions/simple_ngsi10_subscription_exact_city_with_isPattern_and_deeper_service_path.test
>   o  0675_servicePath_in_ngsi10_subscriptions/simple_ngsi10_subscription_with_isPattern_and_deeper_service_path.test
>   o  0675_servicePath_in_ngsi10_subscriptions/simple_ngsi10_subscription_with_recursive_service_path_and_isPattern.test
>   o  0706_direct_https_notifications/direct_https_notifications.test
>   o  0706_direct_https_notifications/direct_https_notifications_no_accept_selfsigned.test
>   o  0876_attribute_dates/attribute_dates.test
>   o  0876_attribute_dates/attrs_dates_overridden_by_user_subs.test
>   o  0876_attribute_dates/filter_attr_datecreated_sub.test
>   o  0876_attribute_dates/filter_attr_datemodified_sub.test
>   o  0876_entity_dates/entity_dates_overridden_by_user_subs.test
>   o  0876_entity_dates/filter_datecreated_sub.test
>   o  0876_entity_dates/filter_datemodified_sub.test
>   o  0943_overnotification_bug/overnotification_bug.test
>   o  0978_filter_attribute_pattern/csub_attribute_pattern.test
>   o  0978_filter_attribute_pattern/csub_attribute_pattern_nocache.test
>   o  1048_subscription_cache/csub_update_with_notif.test
>   o  1129_strings_in_q_uri_param/q_query_strings_for_subscriptions.test
>   o  1156_mq_as_uri_param_for_metadata/mq_as_uri_param_for_metadata.test
>   o  1156_mq_as_uri_param_for_metadata/q_and_mq_as_uri_param_for_metadata.test
>   o  1156_mqfilters_and_compounds/mqfilters_and_compounds_deeper.test
>   o  1156_mqfilters_and_compounds/mqfilters_and_compounds_equals_subs.test
>   o  1156_mqfilters_and_compounds/mqfilters_and_compounds_existence_subs.test
>   o  1156_mqfilters_and_compounds/mqfilters_and_compounds_lt_le_gt_ge_subs.test
>   o  1156_mqfilters_and_compounds/mqfilters_and_compounds_patternmatch_subs.test
>   o  1156_qfilters_and_compounds/qfilters_and_compounds_deeper.test
>   o  1156_qfilters_and_compounds/qfilters_and_compounds_equals_subs.test
>   o  1156_qfilters_and_compounds/qfilters_and_compounds_existence_subs.test
>   o  1156_qfilters_and_compounds/qfilters_and_compounds_lt_le_gt_ge_subs.test
>   o  1156_qfilters_and_compounds/qfilters_and_compounds_patternmatch_subs.test
>   o  1308_subscription_cache/cache_reset_from_db.test
>   o  1308_subscription_cache/subCacheThrottlingAndExpiration.test
>   o  1308_subscription_cache/subCacheThrottlingAndExpirationAfterSubUpdate.test
>   o  1308_subscription_cache/subCacheThrottlingAndExpirationAfterSubUpdateWhenSubExpired.test
>   o  1316_create_subscription/post_subs_expr_q_exist_attr.test
>   o  1316_create_subscription/post_subs_expr_q_not_exist_attr.test
>   o  1328_pause_subscriptions/pause_subscriptions_notif_cache.test
>   o  1328_pause_subscriptions/pause_subscriptions_notif_nocache.test
>   o  1449_notification_mode_cases/onchange_persisent.test
>   o  1449_notification_mode_cases/onchange_simulated.test
>   o  1449_notification_mode_cases/onchange_threadpool.test
>   o  1494_subscription_alteration_types/sub_alteration_type_entity_change.test
>   o  1494_subscription_alteration_types/sub_alteration_type_entity_create.test
>   o  1494_subscription_alteration_types/sub_alteration_type_entity_delete.test
>   o  1494_subscription_alteration_types/sub_alteration_type_entity_delete2.test
>   o  1494_subscription_alteration_types/sub_alteration_type_entity_update.test
>   o  1494_subscription_alteration_types/sub_alteration_type_entity_update3.test
>   o  1678_geosubscriptions/geosubscription_mindistance_zero.test
>   o  1678_geosubscriptions/geosubscriptions.test
>   o  1678_geosubscriptions/geosubscriptions_plus_q_filter.test
>   o  1705_csub_cache_objects/csub_cache_string_filter_update.test
>   o  1898_no_multiservice_and_tenant_used/no_multiservice_and_tenant_used.test
>   o  1903_colon_as_double_equal_synonym/csub_filter_colon_as_enum.test
>   o  1903_colon_as_double_equal_synonym/csub_filter_colon_as_equal.test
>   o  1903_colon_as_double_equal_synonym/csub_filter_colon_as_range.test
>   o  1945_filter_string_comparison/csub_filter_string_greater_or_equal_than.test
>   o  1945_filter_string_comparison/csub_filter_string_greater_than.test
>   o  1945_filter_string_comparison/csub_filter_string_inside_range.test
>   o  1945_filter_string_comparison/csub_filter_string_less_or_equal_than.test
>   o  1945_filter_string_comparison/csub_filter_string_less_than.test
>   o  1945_filter_string_comparison/csub_filter_string_outside_range.test
>   o  2015_notification_templates/notification_templates_many_notifications_after_update.test
>   o  2015_notification_templates/notification_templates_many_queue_statistics.test
>   o  2064_attrs_blacklist_subs/2064_attrs_blacklist_complex.test
>   o  2064_attrs_blacklist_subs/blacklist_after_sub_up.test
>   o  2064_attrs_blacklist_subs/whitelist_after_sub_up.test
>   o  2222_notif_not_sent_on_true/notif_not_sent_on_true.test
>   o  2445_mq_date_processing_subs/mq_date_processing_subs.test
>   o  2507_csub_metadata_field/special_metadata_in_notifications.test
>   o  2507_csub_metadata_field/special_metadata_in_notifications_ngsiv1.test
>   o  2549_append_or_update_in_actionType_in_notifications/append_or_update_in_actionType_in_notifications.test
>   o  2553_wrong_metadata_type_in_prev_value_in_notification/wrong_metadata_type_in_prev_value_in_notification.test
>   o  2588_missing_prevval_on_changed_type/2588_missing_prevval_on_changed_type.test
>   o  2670_milliseconds_for_DateTime_attrs/2670_milliseconds_filter_sub.test
>   o  2750_common_metrics/2750_notifications.test
>   o  2986_apiv2_notifications/apiv2_notification_all_json_types.test
>   o  2986_apiv2_notifications/apiv2_notification_common_geo_json_types.test
>   o  2986_apiv2_notifications/apiv2_notification_reception.test
>   o  2995_q_in_subscriptions/mq_compound_does-not-exist_sub_GTE.test
>   o  2995_q_in_subscriptions/mq_compound_does-not-exist_sub_LTE.test
>   o  2995_q_in_subscriptions/mq_compound_non-compatible-type_sub_GTE.test
>   o  2995_q_in_subscriptions/mq_compound_non-compatible-type_sub_LTE.test
>   o  2995_q_in_subscriptions/mq_value_does-not-exist_sub_GTE.test
>   o  2995_q_in_subscriptions/mq_value_does-not-exist_sub_LTE.test
>   o  2995_q_in_subscriptions/mq_value_non-compatible_sub_GTE.test
>   o  2995_q_in_subscriptions/mq_value_non-compatible_sub_LTE.test
>   o  2995_q_in_subscriptions/q_compound_does-not-exist_sub_GTE.test
>   o  2995_q_in_subscriptions/q_compound_does-not-exist_sub_LTE.test
>   o  2995_q_in_subscriptions/q_compound_non-compatible-type_sub_GTE.test
>   o  2995_q_in_subscriptions/q_compound_non-compatible-type_sub_LTE.test
>   o  2995_q_in_subscriptions/q_value_does-not-exist_sub_GTE.test
>   o  2995_q_in_subscriptions/q_value_does-not-exist_sub_LTE.test
>   o  2995_q_in_subscriptions/q_value_non-compatible_sub_GTE.test
>   o  2995_q_in_subscriptions/q_value_non-compatible_sub_LTE.test
>   o  2998_null_not_working_in_q_and_mq_filter/null_not_working_in_mq_for_subscription.test
>   o  2998_null_not_working_in_q_and_mq_filter/null_not_working_in_q_for_subscription.test
>   o  3001_mqtt/mqtt_notification_custom.test
>   o  3001_mqtt/mqtt_subscription_multibroker.test
>   o  3154_notification_problem_with_custom_header/notification_problem_with_custom_header.test
>   o  3310_preupdate_metadata_content_included_in_notifications/preupdate_metadata_content_included_in_notifications.test
>   o  3533_null_in_datetime_and_geo/null_in_datetime_attr_subs.test
>   o  3533_null_in_datetime_and_geo/null_in_datetime_md_subs.test
>   o  3533_null_in_datetime_and_geo/null_in_geo_box_subs.test
>   o  3533_null_in_datetime_and_geo/null_in_geo_json_subs.test
>   o  3533_null_in_datetime_and_geo/null_in_geo_line_subs.test
>   o  3533_null_in_datetime_and_geo/null_in_geo_point_subs.test
>   o  3533_null_in_datetime_and_geo/null_in_geo_polygon_subs.test
>   o  3906_onlychangedattrs_not_working_with_csub_cache/onlychangedattrs_not_working_with_csub_cache.test
>   o  4069_status_inactive_not_working_with_cache_in_ha/status_inactive_not_working_with_cache_in_ha.test
>   o  4094_posix_regex/posix_regex_idpattern.test
>   o  4120_string_and_null_mixed_filter/sub_string_and_null_mixed_filter.test
> ```
> 
> A huge number of tests are failing... I have had a look to the detail of a couple of them:
> 
> ![imagen](https://user-images.githubusercontent.com/1534240/190465214-6c5f35b5-ebde-45af-9cd3-9f8b6ffb15c5.png)
> 
> ![imagen](https://user-images.githubusercontent.com/1534240/190465256-60518389-318e-49e4-ba22-26753490a932.png)
> 
> So it seems that expected notification are not generated. Probably this is due to the new query has some flaw and it doesn't "capture" the subscriptions associated to the notification that should be triggered according to test script.
> 
> Just in case this feedback may be useful...

Hi @fgalan 

Thanks for these pointers.

Would it be possible just to get the list of documents (from test cases) like we have one in this [mongo playground](https://mongoplayground.net/p/WUY6MsL-D9o)? If I have the list of documents then I can easily replace them with the one in this playground link and it'd be easy to see the documents for which the output is incorrect.

If we do not have the list of documents, then it is convenient to understand what are the cases where the new code is failing.

fgalan commented:
> Would it be possible just to get the list of documents (from test cases) like we have one in this [mongo playground](https://mongoplayground.net/p/WUY6MsL-D9o)? If I have the list of documents then I can easily replace them with the one in this playground link and it'd be easy to see the documents for which the output is incorrect.
> 
> If we do not have the list of documents, then it is convenient to understand what are the cases where the new code is failing.

Good idea :)

Let's focus on 0674_servicepath_in_notifications/service_path_in_single_onchange_notification.test case. It has a pretty simple script:

```
# 01. Add ONCHANGE subscription to E.*/ET with service-path /entity
# 02. Create entity E1/ET with service-path /entity
# 03. Query E.*/ET with service-path /entity
# 04. Query accumulator
```

If the `dbDrop CB` is commented in the last line, then DB keeps the subscription document that should be captured by your new query:

```
> use ftest
switched to db ftest
> db.csubs.find().pretty()
{
	"_id" : ObjectId("633308e274324a7f1d1283eb"),
	"expiration" : NumberLong(1666880994),
	"reference" : "http://127.0.0.1:9997/notify",
	"custom" : false,
	"timeout" : NumberLong(0),
	"throttling" : NumberLong(-1),
	"maxFailsLimit" : NumberLong(-1),
	"servicePath" : "/entity",
	"status" : "active",
	"statusLastChange" : 1664288994.0191789,
	"entities" : [
		{
			"id" : "E.*",
			"isPattern" : "true",
			"type" : "ET",
			"isTypePattern" : false
		}
	],
	"attrs" : [ ],
	"metadata" : [ ],
	"blacklist" : false,
	"onlyChanged" : false,
	"covered" : false,
	"conditions" : [
		"A"
	],
	"expression" : {
		"q" : "",
		"mq" : "",
		"geometry" : "",
		"coords" : "",
		"georel" : ""
	},
	"altTypes" : [ ],
	"format" : "JSON"
}
```

And the triggering entity:

```
> db.entities.find().pretty()
{
	"_id" : {
		"id" : "E1",
		"type" : "ET",
		"servicePath" : "/entity"
	},
	"attrNames" : [
		"A"
	],
	"attrs" : {
		"A" : {
			"type" : "string",
			"creDate" : 1664288994.100778,
			"modDate" : 1664288994.100778,
			"value" : "V",
			"mdNames" : [ ]
		}
	},
	"creDate" : 1664288994.100778,
	"modDate" : 1664288994.100778,
	"lastCorrelator" : "da791ec2-3e70-11ed-9487-000c29583ca5"
}
```

I hope it helps...

anupamme commented:
> 

Hi @fgalan 

I used the document you shared made a [playground link](https://mongoplayground.net/p/HtIXC2H95do) for it.

I made following changes to the document:

1. Added an enclosing array (or [])
2. Replaced NumberLong(0) by 0, NumberLong(-1) by -1 

Now when we run it through the expression it does not qualify because it has only 1 entity and it does not satisfy all the conditions (e.g. type is not "" or "Room").

So I was wondering what was expected out of the expression in this scenario? Which obviously didn't happen that is why a failing test.

Or the expectation is that the document should be parse-able as it is without making the changes I did.

fgalan commented on src/lib/mongoBackend/MongoCommonUpdate.cpp:
```diff
@@ -1280,7 +1281,62 @@ static void fill_idPtypeNP
+
+  map_obj.append("input", "$entities");
+  
+  regex_obj.append("input", "Room1");
```
This should be hardwired. It depends on the entityType. Have a look to L1277

I think it should be:

```suggestion
  regex_obj.append("input", entityId);
```

fgalan commented on src/lib/mongoBackend/MongoCommonUpdate.cpp:
```diff
@@ -1280,7 +1281,62 @@ static void fill_idPtypeNP
+  and_arr.append(and_second.obj());
+
+  eq_arr_2.append("$$this.type");
+  eq_arr_2.append("Room");
```
This should be hardwired. It depends on the entity type. Have a look to L1274

I think it should be:

```suggestion
  eq_arr_2.append(entityType);
```

fgalan commented:
> I used the document you shared made a [playground link](https://mongoplayground.net/p/HtIXC2H95do) for it.
> 
> I made following changes to the document:
> 
>    1. Added an enclosing array (or [])
>    2. Replaced NumberLong(0) by 0, NumberLong(-1) by -1
> 
> Now when we run it through the expression it does not qualify because it has only 1 entity and it does not satisfy all the conditions (e.g. type is not "" or "Room").
> 
> So I was wondering what was expected out of the expression in this scenario? Which obviously didn't happen that is why a failing test.
> 
> Or the expectation is that the document should be parse-able as it is without making the changes I did.

I have provided [a couple of comments](https://github.com/telefonicaid/fiware-orion/pull/4213#pullrequestreview-1126968346) that could help to fix the query although not sure if that's enough...

I'll try to have a deeper look as soon as possible.

fgalan commented:
Let's take a step back...

This is the condition to be replaced by the new query:

```
  bgP->functionIdPtypeNP = std::string("function()") +
         "{" +
            "for (var i=0; i < this."+CSUB_ENTITIES+".length; i++) {" +
                "if (this."+CSUB_ENTITIES+"[i]."+CSUB_ENTITY_ISPATTERN+" == \"true\" && " +
                    "!this."+CSUB_ENTITIES+"[i]."+CSUB_ENTITY_ISTYPEPATTERN+" && " +
                    "(this."+CSUB_ENTITIES+"[i]."+CSUB_ENTITY_TYPE+" == \""+entityType+"\" || " +
                        "this."+CSUB_ENTITIES+"[i]."+CSUB_ENTITY_TYPE+" == \"\" || " +
                        "!(\""+CSUB_ENTITY_TYPE+"\" in this."+CSUB_ENTITIES+"[i])) && " +
                    "\""+entityId+"\".match(this."+CSUB_ENTITIES+"[i]."+CSUB_ENTITY_ID+")) {" +
                    "return true; " +
                "}" +
            "}" +
            "return false; " +
         "}";
```

In natural language it could be expressed as:

* Given a subscription document, for each item within the `entities` array field, the condition is true if:
    * the item `isPattern` field is `"true" AND
    * the item `isTypePattern` field is false AND
    * the item `type` is empty string OR doesn't exist OR is equal to the target type (entityType) AND
    * the item `id` matches the target id (entityId)

Comparing that "script" with your query there are to flaws.

**First**, with regards to

```
            {
              $eq: [
                "$$this.isTypePattern",
                null
              ]
            }
```

it should be

```
            {
              $eq: [
                "$$this.isTypePattern",
                false
              ]
            }
```

Doing so, it seems to work (have a look to [this playgroung](https://mongoplayground.net/p/lOZrXrIFWie)

**Second**, your OR part has only two elements. In order to be equal to the original function, I think your query should include also:

```
                {
                  "$eq": [
                    "$$this.type",
                    null
                  ]
                }
```

(Playground: https://mongoplayground.net/p/IXbi-5MB7xv)

In this case it doesn't have effect, but it should be there.

anupamme commented:
> Let's take a step back...
> 
> This is the condition to be replaced by the new query:
> 
> ```
>   bgP->functionIdPtypeNP = std::string("function()") +
>          "{" +
>             "for (var i=0; i < this."+CSUB_ENTITIES+".length; i++) {" +
>                 "if (this."+CSUB_ENTITIES+"[i]."+CSUB_ENTITY_ISPATTERN+" == \"true\" && " +
>                     "!this."+CSUB_ENTITIES+"[i]."+CSUB_ENTITY_ISTYPEPATTERN+" && " +
>                     "(this."+CSUB_ENTITIES+"[i]."+CSUB_ENTITY_TYPE+" == \""+entityType+"\" || " +
>                         "this."+CSUB_ENTITIES+"[i]."+CSUB_ENTITY_TYPE+" == \"\" || " +
>                         "!(\""+CSUB_ENTITY_TYPE+"\" in this."+CSUB_ENTITIES+"[i])) && " +
>                     "\""+entityId+"\".match(this."+CSUB_ENTITIES+"[i]."+CSUB_ENTITY_ID+")) {" +
>                     "return true; " +
>                 "}" +
>             "}" +
>             "return false; " +
>          "}";
> ```
> 
> In natural language it could be expressed as:
> 
> * Given a subscription document, for each item within the `entities` array field, the condition is true if:
>   
>   * the item `isPattern` field is `"true" AND
>   * the item `isTypePattern` field is false AND
>   * the item `type` is empty string OR doesn't exist OR is equal to the target type (entityType) AND
>   * the item `id` matches the target id (entityId)
> 
> Comparing that "script" with your query there are to flaws.
> 
> **First**, with regards to
> 
> ```
>             {
>               $eq: [
>                 "$$this.isTypePattern",
>                 null
>               ]
>             }
> ```
> 
> it should be
> 
> ```
>             {
>               $eq: [
>                 "$$this.isTypePattern",
>                 false
>               ]
>             }
> ```
> 
> Doing so, it seems to work (have a look to [this playgroung](https://mongoplayground.net/p/lOZrXrIFWie)
> 
> **Second**, your OR part has only two elements. In order to be equal to the original function, I think your query should include also:
> 
> ```
>                 {
>                   "$eq": [
>                     "$$this.type",
>                     null
>                   ]
>                 }
> ```
> 
> (Playground: https://mongoplayground.net/p/IXbi-5MB7xv)
> 
> In this case it doesn't have effect, but it should be there.

I have made these changes. Just to keep in mind I had the value as "missing" not null as per the [SO answer](https://stackoverflow.com/questions/73624483/embed-operator-exists-within-the-operator-map)

fgalan commented on src/lib/mongoBackend/MongoCommonUpdate.cpp:
```diff
@@ -1280,7 +1281,69 @@ static void fill_idPtypeNP
+  and_arr.append(and_third.obj());
+
+  type_obj.append("$type", "$$this.isTypePattern");
+  eq_arr_4.append(type_obj.obj());
+  eq_arr_4.append(false);
```
Not sure if I'm missing something... but it seems that eq_arr_4 is never appended to the holding element...

fgalan commented on src/lib/mongoBackend/MongoCommonUpdate.cpp:
```diff
@@ -1280,7 +1281,69 @@ static void fill_idPtypeNP
+  or_arr.append(eq_obj_2.obj());
+  
+  eq_arr_5.append("$$this.type");
+  eq_arr_5.append(NULL);
```
It seems this is causing a compilation error:

```
/opt/fiware-orion/src/lib/mongoBackend/MongoCommonUpdate.cpp: In function 'void fill_idPtypeNP(CSubQueryGroup*, const string&, const string&, const string&, const string&)':
/opt/fiware-orion/src/lib/mongoBackend/MongoCommonUpdate.cpp:1330:23: error: call of overloaded 'append(NULL)' is ambiguous
 1330 |   eq_arr_5.append(NULL);
      |                       ^
In file included from /opt/fiware-orion/src/lib/mongoBackend/compoundValueBson.h:32,
                 from /opt/fiware-orion/src/lib/mongoBackend/MongoCommonUpdate.cpp:59:
/opt/fiware-orion/src/lib/mongoDriver/BSONArrayBuilder.h:56:8: note: candidate: 'void orion::BSONArrayBuilder::append(const string&)'
   56 |   void append(const std::string& value);
      |        ^~~~~~
/opt/fiware-orion/src/lib/mongoDriver/BSONArrayBuilder.h:57:8: note: candidate: 'void orion::BSONArrayBuilder::append(const char*)'
   57 |   void append(const char* value);
      |        ^~~~~~
/opt/fiware-orion/src/lib/mongoDriver/BSONArrayBuilder.h:58:8: note: candidate: 'void orion::BSONArrayBuilder::append(double)'
   58 |   void append(double value);
      |        ^~~~~~
/opt/fiware-orion/src/lib/mongoDriver/BSONArrayBuilder.h:59:8: note: candidate: 'void orion::BSONArrayBuilder::append(bool)'
   59 |   void append(bool value);
      |        ^~~~~~
```

I think you should use:

```suggestion
  eq_arr_5.appendNull();
```

anupamme replied:
fixed.

fgalan commented:
After the changes in https://github.com/telefonicaid/fiware-orion/compare/111147d...188153e things have improved a bit. For instance, the test 0674_servicepath_in_notifications/service_path_in_single_onchange_notification.test now works.

However, tests are still failing (for instance: 1898_no_multiservice_and_tenant_used/no_multiservice_and_tenant_used.test. I think a couple of changes should be donde, aligned with the [SOF Q&A](https://stackoverflow.com/questions/73624483/embed-operator-exists-within-the-operator-map).

In particular, from the last playground I suggested https://mongoplayground.net/p/IXbi-5MB7xv:

One: change

```
              $eq: [
                "$$this.isTypePattern",
                false
              ]
```

to

```
              $or: [
                {
                  $eq: [
                    "$$this.isTypePattern",
                    false
                  ]
                },
                {
                  $eq: [
                    {
                      $type: "$$this.isTypePattern"
                    },
                    "missing"
                  ]
                }
              ]
```

This captures two situations: isTypePattern is there and its value is `false` or isTypePattern is not there (which implicetely also means `false`).

Two: change

```
                {
                  "$eq": [
                    "$$this.type",
                    null
                  ]
                }
```
to

```
                  "$eq": [
                    {
                      $type: "$$this.type"
                    },
                    "missing"
                  ]
```

Which is the right way to detect a field is missing, according to [SOF Q&A](https://stackoverflow.com/questions/73624483/embed-operator-exists-within-the-operator-map).

Example playgrounds:

* Modified query applied to the document in 0674_servicepath_in_notifications/service_path_in_single_onchange_notification.test: https://mongoplayground.net/p/9GVSiERpUAc
* Modified query applied to the document in 1898_no_multiservice_and_tenant_used/no_multiservice_and_tenant_used.test: https://mongoplayground.net/p/PlLo8KZCW3J

anupamme commented:
hi @fgalan 

The functional tests are now passing and I have checked in changes for style changes as well.

Should I go ahead and replicate the other 3 occurrences of $where by writing the similar code? Or there are any other code review comments I should take care of?

fgalan commented:
> The functional tests are now passing and I have checked in changes for style changes as well.

Great! The work is progressing :)

> Should I go ahead and replicate the other 3 occurrences of $where by writing the similar code? Or there are any other code review comments I should take care of?

Please go ahead with the 3 other functions. I'd suggest to use the same approach:

1. Try to model in natural language the current $where expression
2. Model it with a $expr query
3. Modify the code to implement that and check if functional tests are still passing

I'd suggest doing it step by step. I mean, next in this PR would be a commit for one of the three pending functions and, once we check test are ok, continue with the next one.

anupamme commented:
Hi @fgalan 

I'm having some trouble in writing the $where expressions in natural language. I tried doing it for the function `fill_idNPtypeP`, here is what I came up with:

A loop (for) iterates on the entities field. A positive match (return true) happens if at least one element in the array accomplish the following:
	◦	isPattern is false (true is replaced by false)
	◦	it has a isTypePattern field (nothing to change)
	◦	id == value of variable `entityId`
	◦	type regex matches value of variable `entityType`

But I do not know what will be the values of the variables `entityId` and `entityType`? For example, in earlier case it was Room and Room1.

**Ask:** I think I will run into similar issues in the other 2 cases as well. So would it be possible to give me the conditions in natural language for all the three cases (remaining).

fgalan commented:
> **Ask:** I think I will run into similar issues in the other 2 cases as well. So would it be possible to give me the conditions in natural language for all the three cases (remaining).

Ok, let's try :)

*idNPtypeNP (id no pattern, type no pattern):*

This doesn't uses where, so it doesn't need adaptation.

*idNPtypeP (id no pattern, type pattern)*

Current code:

```
      "{" +
         "for (var i=0; i < this."+CSUB_ENTITIES+".length; i++) {" +
             "if (this."+CSUB_ENTITIES+"[i]."+CSUB_ENTITY_ISPATTERN+" == \"false\" && " +
                 "this."+CSUB_ENTITIES+"[i]."+CSUB_ENTITY_ISTYPEPATTERN+" && " +
                 "this."+CSUB_ENTITIES+"[i]."+CSUB_ENTITY_ID+" == \""+entityId+"\" && " +
                 "\""+entityType+"\".match(this."+CSUB_ENTITIES+"[i]."+CSUB_ENTITY_TYPE+")) {" +
                 "return true; " +
             "}" +
         "}" +
         "return false; " +
      "}";
```

Translation to natural language:

* Given a subscription document, for each item within the `entities` array field, the condition is true if:
    * the item `isPattern` field is `"false"` AND
    * the item `isTypePattern` field is true AND
    * the item `id` field is equal to target id (entityId) AND
    * the item `type` matches the target entity type (entityType)

* idPtypeP (id pattern, type pattern)*

Current code:

```
      "{" +
         "for (var i=0; i < this."+CSUB_ENTITIES+".length; i++) {" +
             "if (this."+CSUB_ENTITIES+"[i]."+CSUB_ENTITY_ISPATTERN+" == \"true\" && " +
                 "this."+CSUB_ENTITIES+"[i]."+CSUB_ENTITY_ISTYPEPATTERN+" && " +
                 "\""+entityId+"\".match(this."+CSUB_ENTITIES+"[i]."+CSUB_ENTITY_ID+") && " +
                 "\""+entityType+"\".match(this."+CSUB_ENTITIES+"[i]."+CSUB_ENTITY_TYPE+")) {" +
                 "return true; " +
             "}" +
         "}" +
         "return false; " +
      "}";
```

Translation to natural language:

* Given a subscription document, for each item within the `entities` array field, the condition is true if:
    * the item `isPattern` field is `"true"` AND
    * the item `isTypePattern` field is true AND
    * the item `id` matches the target entity id (entityId)
    * the item `type` matches the target entity type (entityType)

I hope it helps...

fgalan commented:
> > But I do not know what will be the values of the variables `entityId` and `entityType`? For example, in earlier case it was Room and Room1.

entityId and entityType are the id and type of the target entity, i.e. the entity that comes in the update request and it is checked against the existing subscriptions. Thus, if you update Room1, it will be Room1. If you update E1, it will be E1. It the code, we use variables for that, as you can see.

fgalan commented:
Except for a little error in compliance job, it seems the conversion to query in  fill_idNPtypeP() worked well according functional test cases.

Could you include a link to playground with the sample query? It would be great, for documentation purposes.

Next, you can go ahead with the last case (idPtypeP) :)

anupamme commented:
Hi @fgalan 

I added the code for the function `fill_idPtypeP` and the corresponding query is at the [playground](https://mongoplayground.net/p/HwCnU7BrGHB).

I'm not sure why are the functional tests failing because I do not see any difference in the natural language explanation and the query created. So let me know if you can spot any issues.

fgalan commented:
What about the playground corresponding to fill_idNPtypeP() (the one that works)? Could you include also include it (for completeness and documentation purposes), please?

anupamme commented:
> What about the playground corresponding to fill_idNPtypeP() (the one that works)? Could you include also include it (for completeness and documentation purposes), please?

Here is the link [playground](https://mongoplayground.net/p/rpTUHfJASUf) for the function `fill_idNPtypeP`. 

My bad for not adding it earlier.

fgalan commented on src/lib/mongoBackend/MongoCommonUpdate.cpp:
```diff
@@ -1383,8 +1560,11 @@ static void fill_idPtypeP
   bgP->boPP.appendCode("$where", bgP->functionIdPtypeP);
-
+#else
+  bgP->boPNP.append("$expr", outer_obj.obj());
```
```suggestion
  bgP->boPP.append("$expr", outer_obj.obj());
```

This probably fix the tests that are currently broken.

fgalan commented:
After last fixes, tests are again passing so I think the PR is almost ready! :)

At this point, next steps are:

* Clean-up code modifications, removing the old code (the `#if 0` part)
* Include an entry in the CHANGES_NEXT_RELEASE about the changes in this PR.  Suggestion:

```
- Fix: replace $where operator by $expr in subscription query logic in the -noCache case (#4148)
```

fgalan reviewed (approved):
LGTM. Thanks for your contribution! :)

I'll probably add some code comments to explain the structure of the queries (I think is a valuable piece of knowledge we have found while working on this that shouldn't get lost :). I'll create a PR with that soon.
